File: scripts/check_todos.py
#!/usr/bin/env python3
"""
Check for TODO comments in Solidity files.

When called with file arguments: checks those specific files.
When called with no arguments: checks only git-changed files (modified/added/untracked).
"""

import re
import subprocess
import sys
from pathlib import Path
from typing import List, Tuple

# Pattern to match TODO comments in Solidity
# Matches TODO, FIXME, XXX, HACK in both single-line and multi-line comments
TODO_PATTERN = re.compile(
    r"(//.*\b(todo|fixme|xxx|hack)\b|/\*.*\b(todo|fixme|xxx|hack)\b)",
    re.IGNORECASE
)


def find_todos_in_file(file_path: Path) -> List[Tuple[int, str]]:
    """
    Find TODO comments in a Solidity file.

    Args:
        file_path: Path to the Solidity file

    Returns:
        List of tuples (line_number, line_content) for lines with TODOs
    """
    todos = []
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            for line_num, line in enumerate(f, start=1):
                if TODO_PATTERN.search(line):
                    todos.append((line_num, line.rstrip()))
    except Exception as e:
        print(f"⚠️  Error reading {file_path}: {e}", file=sys.stderr)
    return todos


def get_git_changed_files() -> List[str]:
    """
    Get locally changed Solidity files from git.

    Returns:
        List of changed .sol file paths
    """
    try:
        # Get modified and added files
        result = subprocess.run(
            ['git', 'diff', '--name-only', '--diff-filter=AM', 'HEAD'],
            capture_output=True,
            text=True,
            check=True
        )
        changed_files = [f for f in result.stdout.strip().split('\n') if f.endswith('.sol')]

        # Get untracked files
        result = subprocess.run(
            ['git', 'ls-files', '--others', '--exclude-standard'],
            capture_output=True,
            text=True,
            check=True
        )
        untracked_files = [f for f in result.stdout.strip().split('\n') if f.endswith('.sol')]

        # Combine and filter empty strings
        all_files = [f for f in changed_files + untracked_files if f]
        return all_files
    except subprocess.CalledProcessError:
        return []


def main():
    """Main entry point."""
    # Determine which files to check
    has_file_args = 1 < len(sys.argv)

    if has_file_args:
        # Check specific files passed as arguments
        files_to_check = [f for f in sys.argv[1:] if f.endswith('.sol')]
    else:
        # Check only git-changed files
        files_to_check = get_git_changed_files()
        if not files_to_check:
            print("✅ No locally changed or untracked Solidity files to check for TODO comments.")
            return 0

    if not files_to_check:
        print("✅ No files to check for TODO comments.")
        return 0

    # Check each file for TODOs
    files_checked = 0
    files_with_todos = 0
    total_todos = 0
    todo_found = False

    for file_path_str in files_to_check:
        file_path = Path(file_path_str)

        # Only check if file exists and is a Solidity file
        if not file_path.is_file():
            continue

        files_checked += 1
        todos = find_todos_in_file(file_path)
        if todos:
            if not todo_found:
                print("❌ TODO comments found in Solidity files:")
                print()
                todo_found = True

            files_with_todos += 1
            total_todos += len(todos)

            print(f"📝 {file_path}:")
            for line_num, line in todos:
                print(f"  {line_num}: {line}")
            print()

    # Exit with appropriate message
    file_type = "specified" if has_file_args else "locally changed"
    icon = "❌" if todo_found else "✅"

    print(f"{icon} Found {total_todos} TODO comment(s) in {files_with_todos}/{files_checked} {file_type} Solidity file(s).")

    return 1 if todo_found else 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/check-todos.mjs
#!/usr/bin/env node
/**
 * Check for TODO comments in Solidity files.
 *
 * When called with file arguments: checks those specific files.
 * When called with no arguments: checks only git-changed files (modified/added/untracked).
 */

import { execSync } from 'child_process'
import { existsSync, readFileSync } from 'fs'
import { resolve } from 'path'

// Pattern to match TODO comments in Solidity
// Matches TODO, FIXME, XXX, HACK in both single-line and multi-line comments
const TODO_PATTERN = /\/\/.*\b(todo|fixme|xxx|hack)\b|\/\*.*\b(todo|fixme|xxx|hack)\b/gi

/**
 * Find TODO comments in a Solidity file.
 * @param {string} filePath - Path to the Solidity file
 * @returns {Array<{lineNumber: number, lineContent: string}>} Array of TODO matches
 */
function findTodosInFile(filePath) {
  const todos = []

  try {
    const content = readFileSync(filePath, 'utf-8')
    const lines = content.split('\n')

    lines.forEach((line, index) => {
      if (TODO_PATTERN.test(line)) {
        todos.push({
          lineNumber: index + 1,
          lineContent: line.trimEnd(),
        })
      }
      // Reset regex state for next iteration
      TODO_PATTERN.lastIndex = 0
    })
  } catch (error) {
    console.error(`⚠️  Error reading ${filePath}: ${error}`)
  }

  return todos
}

/**
 * Get locally changed Solidity files from git.
 * @returns {string[]} Array of changed .sol file paths
 */
function getGitChangedFiles() {
  try {
    // Get modified and added files
    const diffOutput = execSync('git diff --name-only --diff-filter=AM HEAD', {
      encoding: 'utf-8',
    }).trim()
    const changedFiles = diffOutput ? diffOutput.split('\n').filter((f) => f.endsWith('.sol')) : []

    // Get untracked files
    const untrackedOutput = execSync('git ls-files --others --exclude-standard', {
      encoding: 'utf-8',
    }).trim()
    const untrackedFiles = untrackedOutput ? untrackedOutput.split('\n').filter((f) => f.endsWith('.sol')) : []

    // Combine and filter empty strings
    return [...changedFiles, ...untrackedFiles].filter((f) => f)
  } catch {
    return []
  }
}

/**
 * Main entry point.
 * @returns {number} Exit code (0 = success, 1 = TODOs found)
 */
function main() {
  // Determine which files to check
  const hasFileArgs = process.argv.length > 2

  let filesToCheck
  if (hasFileArgs) {
    // Check specific files passed as arguments
    filesToCheck = process.argv.slice(2).filter((f) => f.endsWith('.sol'))
  } else {
    // Check only git-changed files
    filesToCheck = getGitChangedFiles()
    if (filesToCheck.length === 0) {
      console.log('✅ No locally changed or untracked Solidity files to check for TODO comments.')
      return 0
    }
  }

  if (filesToCheck.length === 0) {
    console.log('✅ No files to check for TODO comments.')
    return 0
  }

  // Check each file for TODOs
  let filesChecked = 0
  let filesWithTodos = 0
  let totalTodos = 0
  let todoFound = false

  for (const filePathStr of filesToCheck) {
    const filePath = resolve(filePathStr)

    // Only check if file exists and is a Solidity file
    if (!existsSync(filePath)) {
      continue
    }

    filesChecked++
    const todos = findTodosInFile(filePath)

    if (todos.length > 0) {
      if (!todoFound) {
        console.log('❌ TODO comments found in Solidity files:')
        console.log()
        todoFound = true
      }

      filesWithTodos++
      totalTodos += todos.length

      console.log(`📝 ${filePathStr}:`)
      for (const { lineNumber, lineContent } of todos) {
        console.log(`  ${lineNumber}: ${lineContent}`)
      }
      console.log()
    }
  }

  // Exit with appropriate message
  const fileType = hasFileArgs ? 'specified' : 'locally changed'
  const icon = todoFound ? '❌' : '✅'

  console.log(
    `${icon} Found ${totalTodos} TODO comment(s) in ${filesWithTodos}/${filesChecked} ${fileType} Solidity file(s).`,
  )

  return todoFound ? 1 : 0
}

// Run main and exit with appropriate code
process.exit(main())
